File: packages/native/runtime-swift/Sources/PyreonRuntime/PyreonWebView.swift
// PyreonWebView — the native host for the multiplatform `<WebView>`
// primitive. PMTC emits `PyreonWebView(html:)` / `PyreonWebView(src:)`
// for the iOS target; this wraps a `WKWebView` in a SwiftUI view so the
// heavy web-only-rich viz (charts / flow / tables — `@pyreon/charts`,
// `@pyreon/flow`, …) renders inside a native shell. The web target
// renders the same content directly (an `<iframe>`); see
// `@pyreon/primitives`' web `WebView` impl.
//
// ## Policy posture (App Store / Play Store)
//
// The intended use is a HYBRID: a substantial native shell (the canonical
// primitives) with this webview hosting specific heavy-viz screens, NOT a
// thin web wrapper. For the safest review posture, load LOCAL bundled
// assets (`src` resolved against `Bundle.main`) rather than a remote URL —
// that keeps the viz as app content (no remote-code concern), satisfying
// Apple 4.2 / 2.5.2 and Google's webview policy. A remote `src` URL is
// still supported for development / explicitly-online viz.
//
// ## Platform guards
//
// `WKWebView` is in WebKit on both iOS and macOS, but the SwiftUI bridge
// differs: `UIViewRepresentable` (UIKit / iOS) vs `NSViewRepresentable`
// (AppKit / macOS). Both are provided so the package's macOS `swift build`
// AND the iOS app build compile. Same conditional-import pattern as
// PyreonClipboard.

import SwiftUI

#if canImport(WebKit)
import WebKit

public struct PyreonWebView: View {
    private let html: String?
    private let src: String?

    /// `html` — inline HTML to render (e.g. an ECharts page). `src` — a
    /// LOCAL bundled asset name (preferred, policy-safe) or a remote URL.
    /// Supply one; `html` wins if both are set.
    public init(html: String? = nil, src: String? = nil) {
        self.html = html
        self.src = src
    }

    public var body: some View {
        _PyreonWebViewBridge(html: html, src: src)
    }
}

#if canImport(UIKit)
private struct _PyreonWebViewBridge: UIViewRepresentable {
    let html: String?
    let src: String?
    func makeUIView(context: Context) -> WKWebView { WKWebView() }
    func updateUIView(_ webView: WKWebView, context: Context) {
        _loadPyreonWebView(webView, html: html, src: src)
    }
}
#elseif canImport(AppKit)
private struct _PyreonWebViewBridge: NSViewRepresentable {
    let html: String?
    let src: String?
    func makeNSView(context: Context) -> WKWebView { WKWebView() }
    func updateNSView(_ webView: WKWebView, context: Context) {
        _loadPyreonWebView(webView, html: html, src: src)
    }
}
#endif

/// Shared load logic. `html` → `loadHTMLString`; `src` → a bundled local
/// file (resolved against `Bundle.main`, the policy-safe path) falling
/// back to a remote `URL`.
private func _loadPyreonWebView(_ webView: WKWebView, html: String?, src: String?) {
    if let html {
        webView.loadHTMLString(html, baseURL: Bundle.main.bundleURL)
        return
    }
    guard let src else { return }
    if let fileURL = Bundle.main.url(forResource: src, withExtension: nil) {
        webView.loadFileURL(fileURL, allowingReadAccessTo: Bundle.main.bundleURL)
    } else if let url = URL(string: src) {
        webView.load(URLRequest(url: url))
    }
}

#else
// WebKit unavailable (e.g. a headless Linux toolchain) — render nothing
// so the package still compiles. Real targets (iOS / macOS) always have
// WebKit.
public struct PyreonWebView: View {
    public init(html: String? = nil, src: String? = nil) {}
    public var body: some View { EmptyView() }
}
#endif

File: packages/native/runtime-kotlin/src/main/kotlin/com/pyreon/runtime/PyreonWebView.kt
// PyreonWebView — the Android host for the multiplatform `<WebView>`
// primitive (mirror of PyreonWebView.swift). PMTC emits
// `PyreonWebView(html = …)` / `PyreonWebView(src = …)` (+ optional
// `data = …`) on the Kotlin target; this wraps an Android `WebView` in a
// Composable via `AndroidView`, so the heavy web-only-rich viz (charts /
// flow / tables) renders inside a Compose native shell. The web target
// renders the same content directly (an `<iframe>`); see
// `@pyreon/primitives`' web impl.
//
// ## The live-data bridge (`data`)
//
// `data` is a JSON string (PMTC emits `PyreonJson.encode(signal)` for
// `<WebView data={signal}>`). On page load AND whenever `data` changes,
// the runtime PUSHES it into the running page via `evaluateJavascript`:
//
//     window.__pyreonData = <json>;
//     window.dispatchEvent(new Event("pyreondata"));
//
// The hosted page reads `window.__pyreonData` (and re-reads on the
// `pyreondata` event). This is a PUSH into the ALREADY-LOADED page — a
// `data`-only change does NOT reload the WebView, so the chart updates in
// place (no flicker, animation preserved). The page reloads ONLY when
// `html`/`src` changes.
//
// ## Policy posture (Play Store / App Store)
//
// Intended for HYBRID apps — a substantial native Compose shell with this
// WebView hosting specific heavy-viz screens, NOT a thin web wrapper. For
// the safest review posture, ship the viz as a LOCAL asset
// (`file:///android_asset/<src>`) rather than a remote URL.
//
// ## Verification
//
// Not in the runtime package's per-service `kotlinc`-stub verify loop
// (that would need Android-framework + AndroidView stubs); the
// authoritative check is a real `gradle assembleDebug` of a consuming app
// (the device-CI build-proof), where the Android SDK + Compose are present.

package com.pyreon.runtime

import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView

/**
 * Host an Android [WebView] in Compose. Supply [html] (inline HTML) OR
 * [src] (a local `assets/` file name — preferred, policy-safe — or a
 * remote `http(s)` URL); [html] wins if both are set. [data] is an
 * optional JSON string pushed into the page as `window.__pyreonData`
 * (live-updates without reloading; see the file header).
 */
@Composable
fun PyreonWebView(
    html: String? = null,
    src: String? = null,
    data: String? = null,
    modifier: Modifier = Modifier,
) {
    // Per-view bridge state survives recomposition; the WebViewClient
    // (set once in `factory`) closes over it to push on page-load.
    val state = remember { PyreonWebViewState() }
    AndroidView(
        modifier = modifier,
        factory = { context ->
            WebView(context).apply {
                @Suppress("SetJavaScriptEnabled")
                settings.javaScriptEnabled = true
                webViewClient = object : WebViewClient() {
                    override fun onPageFinished(view: WebView, url: String?) {
                        state.loaded = true
                        pushPyreonData(view, state.latestData)
                    }
                }
            }
        },
        update = { webView ->
            state.latestData = data
            val key = (html ?: "") + "" + (src ?: "")
            if (state.loadedKey != key) {
                // html/src changed → (re)load; onPageFinished pushes data.
                state.loadedKey = key
                state.loaded = false
                loadPyreonWebView(webView, html, src)
            } else if (state.loaded) {
                // data-only change → push without reloading.
                pushPyreonData(webView, data)
            }
        },
    )
}

private class PyreonWebViewState {
    var latestData: String? = null
    var loaded: Boolean = false
    var loadedKey: String? = null
}

/** Push the latest JSON into `window.__pyreonData` + fire `pyreondata`. */
private fun pushPyreonData(webView: WebView, json: String?) {
    if (json.isNullOrEmpty()) return
    webView.evaluateJavascript(
        "window.__pyreonData = $json; window.dispatchEvent(new Event(\"pyreondata\"));",
        null,
    )
}

private fun loadPyreonWebView(webView: WebView, html: String?, src: String?) {
    when {
        html != null ->
            webView.loadDataWithBaseURL("file:///android_asset/", html, "text/html", "UTF-8", null)
        src != null && (src.startsWith("http://") || src.startsWith("https://")) ->
            webView.loadUrl(src)
        src != null ->
            webView.loadUrl("file:///android_asset/$src")
    }
}
